### ISSUE TITLE
CustomCall: correctly handling Token objects

### ISSUE THREAD

**Issue by PhilipVinc** [google/jax#5707]
Hello,

over at mpi4jax we are seeing failures with the message
```python
RuntimeError: Internal: Unexpected shape kind for %after-all.2 = token[] after-all(), metadata={op_type="allreduce_mpi" op_name="allreduce_mpi[ comm=<mpi4jax.utils.HashableMPIType object at 0x7fd1025dfa20>\n               op=<mpi4jax.utils.HashableMPIType object at 0x7fd1025df9b0>\n               transpose=False ]"} and shape index {}
```

I believe it is because our CustomCall implementation of MPI primitives, which takes tokens to avoid Jax/XLA reordering mpi operations, does not correctly handle the token objects.
Right now, we do not do nothing (so we do not populate the output token with whatever data it should contain).

What exactly is the C structure of a token, and what should I do or a CustomCall operation accepting a token and returning a token? 

See for example [here](https://github.com/PhilipVinc/mpi4jax/blob/b4a425290e6f006ada5ea422e6f3182694981747/mpi4jax/cython/mpi_xla_bridge_cpu.pyx#L83)

**Comment by hawkinsp** [comment]
Thanks, I forwarded this bug onto the XLA folks.

I suspect that your code is correct. Tokens shouldn't have any real runtime representation; they exist to force compile-time ordering.

**Comment by PhilipVinc** [comment]
By the way, this change is somewhat recent. I believed it appeared approximately in jaxlib 0.1.58.
Also, the error is triggered only spuriously on the CPU backend, but it is always triggered on the GPU(NVIDIA) backend.

**Comment by hawkinsp** [comment]
Yes, my guess is that this is an unintended consequence of some recent refactoring work inside the XLA compiler to use MLIR infrastructure in more places. I think we must not have any tests for this...

**Comment by PhilipVinc** [comment]
@hawkinsp do you have any update on this? It's breaking a big chunk of our workflow.

**Comment by gnecula** [comment]
The XLA team is looking into this.

**Comment by PhilipVinc** [comment]
On jax==0.2.9 and libjax==0.1.61 with CUDA I see the following failure on a very simple code.
Is it related to the same underlying issue or should I file another issue?

```python
>>> import jax
>>> jax.jit(jax.lax.create_token)(1.0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
jax._src.traceback_util.FilteredStackTrace: RuntimeError: Unimplemented: Unimplemented primitive type TOKEN

The stack trace above excludes JAX-internal frames.
The following is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/filippovicentini/Documents/pythonenvs/mpi4jax_env/lib64/python3.8/site-packages/jax/_src/traceback_util.py", line 139, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/home/filippovicentini/Documents/pythonenvs/mpi4jax_env/lib64/python3.8/site-packages/jax/api.py", line 398, in f_jitted
    return cpp_jitted_f(context, *args, **kwargs)
  File "/home/filippovicentini/Documents/pythonenvs/mpi4jax_env/lib64/python3.8/site-packages/jax/api.py", line 289, in cache_miss
    out_flat = xla.xla_call(
  File "/home/filippovicentini/Documents/pythonenvs/mpi4jax_env/lib64/python3.8/site-packages/jax/core.py", line 1275, in bind
    return call_bind(self, fun, *args, **params)
  File "/home/filippovicentini/Documents/pythonenvs/mpi4jax_env/lib64/python3.8/site-packages/jax/core.py", line 1266, in call_bind
    outs = primitive.process(top_trace, fun, tracers, params)
  File "/home/filippovicentini/Documents/pythonenvs/mpi4jax_env/lib64/python3.8/site-packages/jax/core.py", line 1278, in process
    return trace.process_call(self, fun, tracers, params)
  File "/home/filippovicentini/Documents/pythonenvs/mpi4jax_env/lib64/python3.8/site-packages/jax/core.py", line 631, in process_call
    return primitive.impl(f, *tracers, **params)
  File "/home/filippovicentini/Documents/pythonenvs/mpi4jax_env/lib64/python3.8/site-packages/jax/interpreters/xla.py", line 580, in _xla_call_impl
    compiled_fun = _xla_callable(fun, device, backend, name, donated_invars,
  File "/home/filippovicentini/Documents/pythonenvs/mpi4jax_env/lib64/python3.8/site-packages/jax/linear_util.py", line 260, in memoized_fun
    ans = call(fun, *args)
  File "/home/filippovicentini/Documents/pythonenvs/mpi4jax_env/lib64/python3.8/site-packages/jax/interpreters/xla.py", line 714, in _xla_callable
    donated_invars = set_up_aliases(c, xla_args, out_tuple, donated_invars, tuple_args)
  File "/home/filippovicentini/Documents/pythonenvs/mpi4jax_env/lib64/python3.8/site-packages/jax/interpreters/xla.py", line 753, in set_up_aliases
    key = (element.dimensions(), element.numpy_dtype())
RuntimeError: Unimplemented: Unimplemented primitive type TOKEN
```

**Comment by hawkinsp** [comment]
The XLA team tells me the original issue should be fixed using an up to date version of XLA. Can you try building jaxlib from source with an up-to-date XLA (you may need to update the `WORKSPACE` in the JAX tree) and verifying it fixes your original problem?

The new report may or may not be the same problem. Can you share instructions to reproduce it?

**Comment by PhilipVinc** [comment]
How do I select XLA-master? Do I just follow the instructions at https://jax.readthedocs.io/en/latest/developer.html ? Is the `WORKSPACE` documented somewhere?

There's a one-liner that triggers the error in the issue above:
```python
>>> import jax
>>> jax.jit(jax.lax.create_token)(1.0)
```
I'm running on Fedora with CUDA11.1 and jaxlib==0.1.61+cuda111 . The code works fine if I only enable CPU device.

**Comment by dionhaefner** [comment]
@PhilipVinc FWIW this has always been like this. You cannot return token objects from jitted GPU functions. Whether this is intended or not, I don't know.

**Comment by hawkinsp** [comment]
You edit the `WORKSPACE` (see the comments in that file). But now I look, it's actually sufficiently new already, so just building jaxlib from source without changing anything should work. Let me know!

**Comment by hawkinsp** [comment]
@dionhaefner We can probably fix that! But I think it's a separate issue to the one originally raised here...

**Comment by dionhaefner** [comment]
@PhilipVinc I can test the fix, I have everything setup already.
@hawkinsp  I'll report back in an hour or so when JAX has finished building :wink:

**Comment by PhilipVinc** [comment]
I'll open another issue in the meantime

**Comment by dionhaefner** [comment]
:confused: I promised too much, the build failed with

```
ERROR: /lustre/hpc/ocean/dhaefner/.cache/bazel/_bazel_dhaefner/b85bae3607c9b89f9420bdc10d6de09a/external/org_tensorflow/tensorflow/core/common_runtime/BUILD:1647:16: Illegal ambiguous match on configurable attribute "deps" in @org_tensorflow//tensorflow/core/common_runtime:core_cpu_internal:
@org_tensorflow//tensorflow:no_hdfs_support
@org_tensorflow//tensorflow:with_tpu_support
Multiple matches are not allowed unless one is unambiguously more specialized.
ERROR: Analysis of target '//build:build_wheel' failed; build aborted: /lustre/hpc/ocean/dhaefner/.cache/bazel/_bazel_dhaefner/b85bae3607c9b89f9420bdc10d6de09a/external/org_tensorflow/tensorflow/core/common_runtime/BUILD:1647:16: Illegal ambiguous match on configurable attribute "deps" in @org_tensorflow//tensorflow/core/common_runtime:core_cpu_internal:
@org_tensorflow//tensorflow:no_hdfs_support
@org_tensorflow//tensorflow:with_tpu_support
Multiple matches are not allowed unless one is unambiguously more specialized.
```

I can build the latest release with the same setup just fine.

**Comment by dionhaefner** [comment]
Building JAX master worked after doing

```patch
index a09e4ed..2845c53 100755
--- a/build/build.py
+++ b/build/build.py
@@ -273,7 +273,7 @@ build:short_logs --output_filter=DONT_MATCH_ANYTHING
 build:linux --copt=-Wno-stringop-truncation

-# Build with Cloud TPU support.
-build --define=with_tpu_support=true
```

and I can also report that this issue seems to be resolved 🎉 

Good job to the XLA folks!

**Comment by PhilipVinc** [comment]
Resolved in jaxlib 0.1.62
